src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,745 @@
+    return option_ref<decl>(lean_ir_find_env_decl(env.to_obj_arg(), n.to_obj_arg()));
+}
+
+static string_ref * g_mangle_prefix = nullptr;
+static string_ref * g_boxed_mangled_suffix = nullptr;
+
+// reuse the compiler's name mangling to compute native symbol names
+extern "C" object * lean_name_mangle(object * n, object * pre);
+string_ref name_mangle(name const & n, string_ref const & pre) {
+    return string_ref(lean_name_mangle(n.to_obj_arg(), pre.to_obj_arg()));
+}
+
+extern "C" object * lean_ir_format_fn_body_head(object * b);
+format format_fn_body_head(fn_body const & b) {
+    return format(lean_ir_format_fn_body_head(b.to_obj_arg()));
+}
+
+/** \pre Very simple debug output of arbitrary objects, should be extended. */
+void print_object(io_state_stream const & ios, object * o) {
+    if (is_scalar(o)) {
+        ios << unbox(o);
+    } else if (o == nullptr) {
+        ios << "0x0"; // confusingly printed as "0" by the default operator<<
+    } else {
+        // merely following the trace of object addresses is surprisingly helpful for debugging
+        ios.get_stream() << o;
+    }
+}
+
+class interpreter {
+    // stack of IR variable slots
+    std::vector<object *> m_arg_stack;
+    // stack of join points
+    std::vector<fn_body const *> m_jp_stack;
+    struct frame {
+        name m_fn;
+        // base pointers into the stack above
+        size_t m_arg_bp;
+        size_t m_jp_bp;
+    };
+    std::vector<frame> m_call_stack;
+    environment const & m_env;
+    // caches values of nullary functions ("constants")
+    name_map<object *> m_constant_cache;
+    struct symbol_cache_entry {
+        // symbol address; `nullptr` if function does not have native code
+        void * m_addr;
+        // true iff we chose the boxed version of a function where the IR uses the unboxed version
+        bool m_boxed;
+    };
+    // caches symbol lookup successes _and_ failures
+    name_map<symbol_cache_entry> m_symbol_cache;
+
+    /** \brief Get current stack frame */
+    inline frame & get_frame() {
+        return m_call_stack.back();
+    }
+
+    /** \brief Get reference to stack slot of IR variable */
+    inline object * & var(var_id const & v) {
+        // variables are 1-indexed
+        size_t i = get_frame().m_arg_bp + v.get_small_value() - 1;
+        // we don't know the frame size (unless we do an additional IR pass), so we extend it dynamically
+        if (i >= m_arg_stack.size()) {
+            m_arg_stack.resize(i + 1);
+        }
+        return m_arg_stack[i];
+    }
+
+    object * eval_arg(arg const & a) {
+        // an "irrelevant" argument is type- or proof-erased; we can use an arbitrary value for it
+        return arg_is_irrelevant(a) ? box(0) : var(arg_var_id(a));
+    }
+
+    /** \brief Allocate constructor object with given tag and arguments */
+    object * alloc_ctor(ctor_info const & i, array_ref<arg> const & args) {
+        size_t tag = ctor_info_tag(i).get_small_value();
```

Comment:
> Here and in many other places I assumed that all indices and offsets used in the IR will fit in a small, i.e. 31/63-bit, value.

Most of these assumptions are correct, and I also use them all over the place. However, we have to make sure each one of them is correct. For example. the runtime uses a 8-bit value tag. So, it is fine here IF we assume the compiler is rejecting inductive datatypes containing too many constructors.

> Actually, get_small_value returns just an unsigned, which looks like a mistake?

Yes, this looks like a leftover before the transition from `box/unbox` to `size_t`. Recall that at one point the runtime was boxing using `unsigned`, and you and I switched to `size_t`, and tried to ensure that the whole codebase was using `size_t`.

> What was the reason that the IR uses Nat instead of a fixed-size type, ergonomics?

No, it is not just ergonomics. I tried to avoid runtime design decisions from leaking into the high-level objects defined in Lean. `Array` is a good example, in principle, we could use `USize` instead of `Nat`, but it would create problems for people trying to reason about arrays. Their theorems would have to contain artificial assumptions about the `Array` size. Returning to the IR, suppose someone writes a translator from Lean `Expr` to `IR.Expr` and wants to prove some properties about it. If we don't use `Nat`, then they would have to once again add artificial assumptions stating which kind of Lean object can be translated into `IR`. I say these assumptions are artificial because they reflect design decisions of **our** runtime implemented in C/C++. Nothing prevents someone from implementing a different runtime with different characteristics.